src/Stack/Build/Source.hs

```diff
@@ -20,121 +21,190 @@ import qualified    Pantry.SHA256 as SHA256
-                       , LoadedSnapshot
-                       , [LocalPackage] -- FIXME do we really want this? it's in the SourceMap
-                       , Set PackageName -- non-project targets
-                       , SourceMap
-                       )
-loadSourceMapFull needTargets boptsCli = do
-    bconfig <- view buildConfigL
-    (ls, localDeps, targets) <- parseTargets needTargets boptsCli
-    packages <- view $ buildConfigL.to bcPackages
-    locals <- mapM (loadLocalPackage True boptsCli targets) $ Map.toList packages
-    checkFlagsUsed boptsCli locals localDeps (lsPackages ls)
-    checkComponentsBuildable locals
-
-    -- TODO for extra sanity, confirm that the targets we threw away are all TargetAll
-    let nonProjectTargets = Map.keysSet targets `Set.difference` Map.keysSet packages
-
-    -- Combine the local packages, extra-deps, and LoadedSnapshot into
-    -- one unified source map.
-    let goLPI loc n lpi = do
-          let configOpts = getGhcOptions bconfig boptsCli n False False
-          case lpiLocation lpi of
-            -- NOTE: configOpts includes lpiGhcOptions for now, this may get refactored soon
-            PLImmutable pkgloc -> do
-              ident <- getPackageLocationIdent pkgloc
-              return $ PSRemote loc (lpiFlags lpi) configOpts pkgloc ident
-            PLMutable dir -> do -- FIXME this is not correct, we don't want to treat all Mutable as local
-              pp <- mkProjectPackage YesPrintWarnings dir
-              lp' <- loadLocalPackage False boptsCli targets (n, pp)
-              return $ PSFilePath lp' loc
-    sourceMap' <- Map.unions <$> sequence
-      [ return $ Map.fromList $ map (\lp' -> (packageName $ lpPackage lp', PSFilePath lp' Local)) locals
-      , sequence $ Map.mapWithKey (goLPI Local) localDeps
-      , sequence $ Map.mapWithKey (goLPI Snap) (lsPackages ls)
-      ]
-    let sourceMap = sourceMap'
-            `Map.difference` Map.fromList (map (, ()) (toList wiredInPackages))
-
-    return
-      ( targets
-      , ls
-      , locals
-      , nonProjectTargets
-      , sourceMap
-      )
+-- * Target information is only relevant insofar as it effects the
+--   dependency map. The actual current targets for this build are
+--   irrelevant to the cache mechanism, and can be ignored.
+--
+-- * Make sure things like profiling and haddocks are included in the hash
+--
+hashSourceMapData
+    :: (HasConfig env)
+    => WhichCompiler
+    -> Map PackageName DepPackage
+    -> RIO env SourceMapHash
+hashSourceMapData wc smDeps = do
+    path <- encodeUtf8 . T.pack . toFilePath <$> getCompilerPath wc
+    let compilerExe =
+            case wc of
+                Ghc -> "ghc"
+                Ghcjs -> "ghcjs"
+    info <- BL.toStrict . fst <$> proc compilerExe ["--info"] readProcess_
+    immDeps <-
+        fmap B.concat . forM (Map.elems smDeps) $ depPackageHashableContent
+    return $ SourceMapHash (SHA256.hashBytes $ B.concat [path, info, immDeps])
```

Comment:
I know it's done once, nevertheless it is an expensive operation. It seems unnecessary to call `B.concat` twice. Moreover allocation another huge `ByteString` in order to compute it's hash seems wasteful. I'd suggest removing `fmap B.concat . ` from the previous line and doing: `(SHA256.hashBytesLazy $ BL.fromChunks (path:info:immDeps)` here.